Spark 2.4: Add back rewrite data files action

ajantha-bhat opened the pull request:
#3587 has removed deprecated actions. It removed rewrite data files deprecated actions also.
**But rewrite data files was marked deprecated without having an alternate implementation in spark-2.4.**
Hence now spark-2.4 cannot use rewrite data files feature.

**Added back the rewrite data files action for spark-2.4 and updated its deprecated annotations.**

ajantha-bhat commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkActions.java:
```diff
@@ -44,6 +44,11 @@ public DeleteOrphanFiles deleteOrphanFiles(Table table) {
 
+  @Override
+  public RewriteDataFiles rewriteDataFiles(Table table) {
+    return new BaseRewriteDataFilesSparkAction(spark, table);
```
oh, it is an abstract class!

ajantha-bhat replied:
@RussellSpitzer : I should have something like Spark3BinPackStrategy and Spark3SortStrategy.
Then call that from BaseRewriteDataFilesSparkAction2 ?

ajantha-bhat replied:
But we don't have `SparkDistributionAndOrderingUtil`.

I am confused now. We need to bring back deprecated action for this or block rewrite data files for spark 2.4 or fix it by bringing SparkDistributionAndOrderingUtil and other classes form spark3 [not sure if it works]

ajantha-bhat commented:
cc: @RussellSpitzer , @rdblue , @aokolnychyi

rdblue commented:
Can we revert the change that removed the old rewrite files in 2.4 only? That seems like the safest way to fix the regression ahead of the 0.13.0 release.

ajantha-bhat commented:
> Can we revert the change that removed the old rewrite files in 2.4 only? That seems like the safest way to fix the regression ahead of the 0.13.0 release.

I am also considering the same considering the 0.13.0 release timeline. Next version we can rework on this again may be if required. I will raise a revert PR for it now.

ajantha-bhat commented:
@rdblue , @RussellSpitzer , @jackye1995 , @puneetzaroo : 
Reverted the #3587 as suggested. Spark2.4 deprecated class removal will be handled again in the next version after 0.13.0 release.

ajantha-bhat commented:
After a direct revert, one test case failed [org.apache.iceberg.actions.TestRewriteDataFilesAction > testRewriteDataFilesForLargeFile FAILED]
I have analyzed and found it is from #3666, pushed the modified test case in the new commit.

PR is ready now.

RussellSpitzer commented:
We shouldn't revert this I think, instead we should implement the 2.4 version of the new api.

jackye1995 commented:
Instead of reverting the entire `Actions`, can we only revert the ones we have not provided implementations in Spark24? I think the list is `snapshotTable`, `migrateTable`, `rewriteDataFiles`, which corresponds to `CreateAction`, `SnapshotAction` and `RewriteDataFilesAction`.

@RussellSpitzer I think what you said is definitely the right way to fix this, but I feel we should not try to implement that just for the sake of fixing feature regression across releases. People who actually want to use these actions in Spark24 should do the backport.

Any thoughts? @ajantha-bhat @rdblue

ajantha-bhat commented:
Considering 0.13.0 release timeline @rdblue suggested that we revert this change and I too think it as the right step to do.

@RussellSpitzer : 
> instead we should implement the 2.4 version of the new api.

a) For this we should have similar classes of Spark3BinPackStrategy, Spark3SortStrategy and then we need 2.4 version of 
DistributionAndOrderingUtils etc. May take time for me to analyze and handle. So, may be next version we can handle this.
b) It seems snapshotTable, migrateTable also missing after removing the deprecated actions [as they don't have alternate implementations]

@jackye1995 : 
Instead of reverting whole changes, just brining back  snapshotTable, migrateTable, rewriteDataFiles with Actions package makes sense to me. But we also need to remove deprecated annotations from Actions package I guess as it is not really having alternate implementations.

@RussellSpitzer , @rdblue : Can we conclude on this ?

RussellSpitzer commented:
If we are reverting all the deprecation messages also need to be changed, they currently say the function will be removed in 0.13. 

We also don't need to support Sort in 2.4, it never was supported before in 2.4. I mainly don't want to keep this around because the old api had some odd behaviors, if we bring it back we then have two different apis depending on the Spark version which will do materially different things. Although if we want to maintain that as a long term support I think that could be fine. 

The main issue is many of the 3.0 actions are just very difficult to implement in 2.4 (like snapshot and migrate) which I only originally implemented for Spark 3. https://github.com/apache/iceberg/pull/1525/files

So if we do revert, it probably only makes sense to move back `rewriteDataFiles` as I don't think any of the other actions had implementations

jackye1995 commented:
> So if we do revert, it probably only makes sense to move back `rewriteDataFiles` as I don't think any of the other actions had implementations

Yes totally agree, I mentioned `CreateAction` and `SnapshotAction` because we also don't have alternative implementations for Spark 2.4 for those, but they don't have original implementations either. So that boils down to only the `rewriteDataFiles`.

> For this we should have similar classes of Spark3BinPackStrategy, Spark3SortStrategy and then we need 2.4 version of
DistributionAndOrderingUtils etc.

For the future work, I really don't think we should spend effort in re-implementing the action in 2.4. We can mark that as deprecated and add a note "please move to Spark 3.x for the new action implementation".

As Russell said it does not support sort strategy anyway, so it's only gonna be partial support even if implemented. I think only people who absolutely want that should try to backport, and hopefully this can push people to continuously move forward in Spark versions. That's the best way for us to support new Spark versions in the current architecture, by gradually dropping support of features in old versions, otherwise it would be too much burden for developers to maintain all the Spark versions, especially 2.4 because it's so different from the 3.x versions.

ajantha-bhat commented:
@RussellSpitzer , @jackye1995 : thanks for your inputs.
I have added back the rewrite data files action for spark-2.4 and removed its deprecated annotations. 

PR is ready. Please help for review and merge. Thanks.

jackye1995 reviewed (commented):
Mostly looks good to me. Could you add the `@Deprecated` marker in `Actions` and `RewriteDataFilesAction` class, mention this is retained only for `rewriteDataFiles` for backwards compatibiulity, and also point to the fact that (1) this implementation is no longer maintained, (2) a new implementation is available in Spark 3.x and that is preferred.

jackye1995 reviewed (approved):
looks good to me, @RussellSpitzer could you take another look?

RussellSpitzer commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/Actions.java:
```diff
@@ -0,0 +1,109 @@
+  /**
+   * @deprecated since 0.12.0, Used for supporting {@link RewriteDataFilesAction}
+   * in spark-2.4 for backward compatibility.
+   * Also this implementation is no longer maintained.
```
I would changed this to "This implementation is no longer maintained, the new implementation {link} is available with Spark 3.x"

The structure here is a bit complicated.

RussellSpitzer commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/Actions.java:
```diff
@@ -0,0 +1,109 @@
+   * @deprecated since 0.12.0, Used for supporting {@link RewriteDataFilesAction}
+   * in spark-2.4 for backward compatibility.
+   * Also this implementation is no longer maintained.
+   * Preferred to use new implementation {@link BaseRewriteDataFilesSpark3Action} available with Spark 3.x.
```
Same change for deprecation message here and below

RussellSpitzer commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java:
```diff
@@ -0,0 +1,72 @@
+
+/**
+ * @deprecated since 0.12.0, keeping this in spark-2.4 for backward compatibility.
+ * Also this implementation is no longer maintained.
```
Same Deprecation message comment, did we have this class public before? I just ask because the constructor is not public so we could keep this package private too unless the package structure doesn't allow it

RussellSpitzer replied:
Actually forget that on the "package private" we probably don't want to change the api

RussellSpitzer commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/SparkActions.java:
```diff
@@ -0,0 +1,36 @@
+
+/**
+ * @deprecated since 0.12.0, Used for supporting {@link RewriteDataFilesAction}
+ * in spark-2.4 for backward compatibility.
```
Deprecation message again

RussellSpitzer reviewed (approved):
I just have a few nits to the deprecation message. I just want to make sure the message is clear.

ajantha-bhat commented:
@RussellSpitzer: I have updated the deprecated message as per your suggestion. Thanks.

jackye1995 commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/Actions.java:
```diff
@@ -0,0 +1,109 @@
+public class Actions {
+
+  /*
+  We load the actual implementation of Actions via reflection to allow for differences
```
nit: comment style incorrect

jackye1995 commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/Actions.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * An API for interacting with actions in Spark.
+ *
+ * @deprecated since 0.12.0, Used for supporting {@link RewriteDataFilesAction} in spark-2.4 for backward compatibility.
```
nit: `used` instead of `Used`, `Spark 2.4` instead of `spark-2.4`

jackye1995 commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/Actions.java:
```diff
@@ -0,0 +1,109 @@
+  /**
+   * @deprecated since 0.12.0, Used for supporting {@link RewriteDataFilesAction}
+   * in spark-2.4 for backward compatibility.
+   * This implementation is no longer maintained,
```
nit: can be put in 1 line

jackye1995 commented:
@ajantha-bhat sorry for a bit more nitpicking, please let me know when you fix those, and I will merge this. Thanks for the work!!

jackye1995 commented on spark/v2.4/spark/src/main/java/org/apache/iceberg/actions/Actions.java:
```diff
@@ -0,0 +1,109 @@
+ * An API for interacting with actions in Spark.
+ *
+ * @deprecated since 0.12.0, Used for supporting {@link RewriteDataFilesAction} in spark-2.4 for backward compatibility.
+ * This implementation is no longer maintained,
```
I think we don't need to link to `BaseRewriteDataFilesSpark3Action`, this is not imported class anyway. We can just say `This implementation is no longer maintained, the new implementation is available with Spark 3.x`. It also puts the whole sentence to the same line.

ajantha-bhat commented:
@jackye1995 : Done. I have handled the new comments. Thanks
